File: android/app/src/main/java/edu/gatech/cs2340/rajsjarfiles/spacetrader/entity/universe/ResourceClassification.java
package edu.gatech.cs2340.rajsjarfiles.spacetrader.entity.universe;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Random;

/**
 * Represents a planet's resource classification.
 */
public enum ResourceClassification {

    NO_SPECIAL_RESOURCES(35),

    MINERAL_RICH(10),
    MINERAL_POOR(10),

    LOTS_OF_WATER(10),

    RICH_SOIL(10),
    POOR_SOIL(10),

    RICH_FAUNA(10),

    WEIRD_MUSHROOMS(10),

    LOTS_OF_HERBS(10),

    ARTISTIC(10),
    WARLIKE(30),
    LIFELESS(5),
    DESERT(10);

    //ARTISTIC,
    //WARLIKE,

    // Probability of having certain resources in a planet
    private int probOfHavingResource;

    /**
     * Constructor for resource classification enum class
     *
     * @param prob probability of having this resource in a planet
     */
    ResourceClassification(int prob) {
        this.probOfHavingResource = prob;
    }

    /**
     * Set probability of resource to exists in a planet
     *
     * @param prob probability of resource to exists
     */
    public void setProbability(int prob) {
        this.probOfHavingResource = prob;
    }

    /**
     * Get random resource classification for certain habitats.
     * @param habitats habitat of a planet where the resource is allocated
     * @return single random resource classification
     */
    public static ResourceClassification getRandomResourceClass(
            Habitats habitats) {
        Random rand = new Random();

        // List that store possible resource classification
        Collection<ResourceClassification> pr = new ArrayList<>();

        // Filter out resources that are possible to exists in this habitat
        for (ResourceClassification resource
                : ResourceClassification.values()) {
            if ((resource != null)
                    && !habitats.getImpossibleResources().contains(resource)) {
                pr.add(resource);
            }
        }

        // Randomize the choice but taking care of the probability
        ArrayList<Integer> distributionResource = new ArrayList<>();
        for (ResourceClassification r : pr) {
            for (int i = 0; i < r.probOfHavingResource; i++) {
                distributionResource.add(r.ordinal());
            }
        }
        int randNum = distributionResource.get(
                rand.nextInt(distributionResource.size()));
        return ResourceClassification.values()[randNum];
    }
}

File: ios/SpaceTrader/SpaceTrader/src/entity/universe/ResourceClassification.swift
//
//  .swift
//  SpaceTrader
//
//  Created by Thomas Suarez on 4/20/19.
//  Copyright © 2019 RajsJarFiles. All rights reserved.
//

import Foundation

struct ResourceClassification : CustomStringConvertible{
    
    public static let NO_SPECIAL_RESOURCES = ResourceClassification(35, "No Special Resources")
    public static let MINERAL_RICH = ResourceClassification(10, "Rich Minerals")
    public static let MINERAL_POOR = ResourceClassification(10, "Poor Minerals")
    public static let LOTS_OF_WATER = ResourceClassification(10, "Lots of Water")
    public static let RICH_SOIL = ResourceClassification(10, "Rich Soil")
    public static let POOR_SOIL = ResourceClassification(10, "Poor Soil")
    public static let RICH_FAUNA = ResourceClassification(10, "Rich Fauna")
    public static let WEIRD_MUSHROOMS = ResourceClassification(10, "Weird Mushrooms")
    public static let LOTS_OF_HERBS = ResourceClassification(10, "Lots of Herbs")
    public static let ARTISTIC = ResourceClassification(10, "Artistic")
    public static let WARLIKE = ResourceClassification(30, "Warlike")
    public static let LIFELESS = ResourceClassification(5, "Lifeless")
    public static let DESERT = ResourceClassification(10, "Desert")
    
    public static let values = [
        NO_SPECIAL_RESOURCES,
        MINERAL_RICH,
        MINERAL_POOR,
        LOTS_OF_WATER,
        RICH_SOIL,
        POOR_SOIL,
        RICH_FAUNA,
        WEIRD_MUSHROOMS,
        LOTS_OF_HERBS,
        ARTISTIC,
        WARLIKE,
        LIFELESS,
        DESERT
    ]
    
    var probOfHavingResource : Int
    var name : String
    
    var description: String {
        get {
            return self.name
        }
    }
    
    init(_ prob : Int, _ name : String) {
        self.probOfHavingResource = prob
        self.name = name
    }
    
    static func getRandomResourceClass() -> ResourceClassification {
        let r = Int.random(in: 0..<ResourceClassification.values.count)
        return ResourceClassification.values[r]
    }
    
}
